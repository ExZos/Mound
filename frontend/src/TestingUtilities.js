import GeneralComponent from './components/GeneralComponent';

class TestingUtilities extends GeneralComponent {
  addUserSessionItem(userID, spaceID, pollID=undefined) {
    const user = {
      id: userID,
      name: 'user' + userID,
      space: spaceID,
      space_name: 'space' + spaceID,
      space_status: true,
      poll: pollID
    };

    this.addToSessionArrayItem('users', user);
  }
}

export default TestingUtilities;
