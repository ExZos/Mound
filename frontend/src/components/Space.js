import React from 'react';

import GeneralComponent from './GeneralComponent';
import Header from './Header';
import ErrorBlock from './ErrorBlock';
import MessageSpace from './MessageSpace';
import { server, api } from '../server';
import '../styles/space.css';

class Space extends GeneralComponent {
  constructor(props) {
    super(props);

    this.state = {
      space: {
        id: this.props.location.state.spaceID,
        name: ''
      },
      user: {
        name: ''
      }
    };
  }

  componentDidMount() {
    this.getSpace();
  }

  getSpace = () => {
    server.get(api.spaces + this.props.location.state.spaceID)
      .then(res => this.setState({
        space: res.data
      }));
  }

  getUserInSpaceByName = () => {
    server.get(api.getUserInSpaceByName + this.state.space.id + '/' + this.state.user.name)
      .then(
        (res) => {
          this.addToSessionArrayItem('users', res.data);
          this.setSessionItem('toggleMessages', true);

          this.setState({
            user: res.data
          });
        }
      )
      .catch(
        (err) => this.showError()
      );
  }

  handleUserChange = (e) => {
    let { name, value } = e.target;
    const user = { ...this.state.user, [name]: value };

    this.setState({
      user: user
    });

    this.hideError();
  }

  handleFormSubmit = (e) => {
    e.preventDefault();
  }

  toggleMessages = () => {
    const users = this.getSessionItem('users')

    if(users && users[this.state.space.id]) {
      return (
        <MessageSpace spaceID={this.state.space.id} />
      )
    }

    return (
      <div>
        <form id="getUserInSpaceByName" onSubmit={this.handleFormSubmit}>
          <input type="text" name="name" placeholder="Type a user name..." autoFocus
            value={this.state.user.name}
            onChange={this.handleUserChange}
          />

          <button onClick={this.getUserInSpaceByName}>ENTER</button>
        </form>

        <ErrorBlock message="NO SUCH USER" show={this.state.showError} />
      </div>
    );
  }

  render() {
    return(
      <div id="space">
        <Header />

        <br />

        <div id="spaceName">
          {this.state.space.name}
        </div>

        <br />

        {this.toggleMessages()}
      </div>
    )
  }
}

export default Space;
