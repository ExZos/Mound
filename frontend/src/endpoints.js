export const endpoints = {
  messages: '/api/messages/',
  spaces: '/api/spaces/',
  users: '/api/users/',
  poll: '/api/polls/',
  votes: '/api/votes/',
  getSpaceByName: '/space/getByName/',
  getUserCountInSpaceForUser: '/space/getUserCountForUser/',
  getUserByName: '/user/getByName/',
  getUsersInSpace: '/user/getInSpace/',
  getUserInSpaceByName: '/user/getInSpaceByName/',
  getUsersInSpaceExceptName: '/user/getInSpaceExceptName/',
  createUserNApproveSpace: 'user/createNApproveSpace/',
  getMessagesInSpace: '/message/getInSpace/',
  getPendingPollsInSpace: '/poll/getPendingInSpace/',
  getPendingUnvotedPollsForUser: '/poll/getPendingUnvotedForUser/',
  getPendingJoinPollInSpaceByName: 'poll/getPendingJoinInSpaceByName/',
  getJoinPollResults: '/poll/getJoinResults/',
  createNameRelatedPoll: '/poll/createNameRelated/',
  getVotesForPoll: 'vote/getForPoll/',
  getPositiveVotesForPoll: 'vote/getPositivesForPoll/',
  createVoteNUpdatePoll: 'vote/createNUpdatePoll/',
}
