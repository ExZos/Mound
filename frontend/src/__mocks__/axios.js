import { endpoints as api } from '../endpoints';

const axios = {
  create: () => axios,
  get: (url) => {
    switch(url) {
      case api.getSpaceByName + "Headon's Floor/":
        return Promise.resolve({
          data: {
            id: 1,
            name: "Headon's Floor",
            status: true
          }
        });
        break;

      default:
        return Promise.reject();
    }
  }
};

export default axios;
