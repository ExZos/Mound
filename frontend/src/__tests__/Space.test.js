import React from 'react';
import { shallow, mount } from 'enzyme';
import axios from 'axios';
import { createBrowserHistory } from 'history';

import Space from '../components/Space';
import TestingUtilities from '../TestingUtilities';
import { endpoints as api } from '../endpoints';

const history = createBrowserHistory();
const tutils = new TestingUtilities();

describe('Space', () => {
  beforeEach(() => {
    sessionStorage.clear();
  });

  it('should render correctly with props but no session', () => {
    const props = {
      state: {
        space: {
          id: 1,
          name: 'space1',
          status: true
        }
      }
    };

    const component = shallow(<Space location={props} />);
    const user = component.find('NavItem.user');
    const space = component.find('NavItem.space');
    const close = component.find('NavItem.close');
    const spaceForm = component.find('form#getUserInSpaceByName');

    expect(component).toMatchSnapshot();
    expect(user).toHaveLength(0);
    expect(space).toHaveLength(1);
    expect(space.children().text()).toEqual('space1');
    expect(close).toHaveLength(1);
    expect(spaceForm).toHaveLength(1);
  });

  it('should render MessageSpace with props and session', () => {
    tutils.addUserSessionItem(2, 2, 2);

    const props = {
      state: {
        space: {
          id: 2,
          name: 'space2',
          status: true
        }
      }
    };

    const component = shallow(<Space location={props} />);
    const user = component.find('NavItem.user');
    const space = component.find('NavItem.space');
    const close = component.find('NavItem.close');
    const spaceForm = component.find('form#getUserInSpaceByName');
    const messageSpace = component.find('MessageSpace');

    expect(component).toMatchSnapshot();
    expect(user).toHaveLength(1);
    expect(user.children().text()).toEqual('user2');
    expect(space).toHaveLength(1);
    expect(space.children().text()).toEqual('space2');
    expect(close).toHaveLength(1);
    expect(spaceForm).toHaveLength(0);
    expect(messageSpace).toHaveLength(1);
  });

  it('should render PendingUser with props and session ', () => {
    tutils.addUserSessionItem(undefined, 3, 3);

    const props = {
      state: {
        space: {
          id: 3,
          name: 'space3',
          status: true
        }
      }
    };

    const component = shallow(<Space location={props} />);
    const user = component.find('NavItem.user');
    const space = component.find('NavItem.space');
    const close = component.find('NavItem.close');
    const spaceForm = component.find('form#getUserInSpaceByName');
    const pendingUser = component.find('PendingUser');

    expect(component).toMatchSnapshot();
    expect(user).toHaveLength(1);
    expect(user.children().text()).toEqual('userundefined');
    expect(space).toHaveLength(1);
    expect(space.children().text()).toEqual('space3');
    expect(close).toHaveLength(1);
    expect(spaceForm).toHaveLength(0);
    expect(pendingUser).toHaveLength(1);
  });

  it('should render PendingSpace with props and session', () => {
    tutils.addUserSessionItem(undefined, 3, 3, false);

    const props = {
      state: {
        space: {
          id: 3,
          name: 'space3',
          status: false
        }
      }
    };

    const component = shallow(<Space location={props} />);
    const user = component.find('NavItem.user');
    const space = component.find('NavItem.space');
    const close = component.find('NavItem.close');
    const spaceForm = component.find('form#getUserInSpaceByName');
    const pendingSpace = component.find('PendingSpace');

    expect(component).toMatchSnapshot();
    expect(user).toHaveLength(1);
    expect(user.children().text()).toEqual('userundefined');
    expect(space).toHaveLength(1);
    expect(space.children().text()).toEqual('space3');
    expect(close).toHaveLength(1);
    expect(spaceForm).toHaveLength(0);
    expect(pendingSpace).toHaveLength(1);
  });

  it('should create a state entry equal to the input value', () => {
    const props = {
      state: {
        space: {
          id: 1,
          name: 'space1',
          status: true
        }
      }
    };

    const component = shallow(<Space location={props} />);
    const input = component.find('input');

    input.props().onChange({
      target: {
        name: 'name',
        value: 'user1'
      }
    });

    expect(component.state('user')).toBeDefined();
    expect(component.state('user')).toHaveProperty('name', 'user1');
  });

  it('should create state and session entries on approved space submit', async () => {
    const props = {
      state: {
        space: {
          id: 1,
          name: 'space1',
          status: true
        }
      }
    };

    const component = shallow(<Space location={props} />);
    const button = component.find('button');
    const spy = jest.spyOn(axios, 'get');

    component.setState({
      user: {
        name: 'user1'
      }
    });
    await button.props().onClick();

    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenLastCalledWith(api.getUserInSpaceByName + '1/user1');
    expect(component.state('user')).toBeDefined();
    expect(component.state('user')).toHaveProperty('id', 1);
    expect(tutils.getSessionItem('users')).toBeDefined();
    expect(tutils.getSessionItem('users')[1]).toBeDefined();
  });

  it('should create state and session entries on pending space/user submit', async () => {
    const props = {
      state: {
        space: {
          id: 2,
          name: 'space2',
          status: true
        }
      }
    };

    const component = shallow(<Space location={props} />);
    const button = component.find('button');
    const spy = jest.spyOn(axios, 'get');

    component.setState({
      user: {
        name: 'user2',
        space: 2
      }
    });
    await button.props().onClick();

    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenLastCalledWith(api.getPendingJoinPollInSpaceByName + '2/user2');
    expect(component.state('user')).toBeDefined();
    expect(component.state('user')).not.toHaveProperty('id');
    expect(component.state('poll')).toBeDefined();
    expect(tutils.getSessionItem('users')).toBeDefined();
    expect(tutils.getSessionItem('users')[2]).toBeDefined();
    expect(tutils.getSessionItem('users')[2]).toHaveProperty('poll', 2);
  });

  it('should not create state or session entries on invalid submit and should trigger modal', async () => {
    const props = {
      state: {
        space: {
          id: 1,
          name: 'space1',
          status: true
        }
      }
    };

    const component = shallow(<Space location={props} />);
    const button = component.find('button');
    const spy = jest.spyOn(axios, 'get');

    component.setState({
      user: {
        name: 'Missing User'
      }
    });
    await button.props().onClick();

    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenLastCalledWith(api.getPendingJoinPollInSpaceByName + '1/Missing User');
    expect(component.state('user')).toBeDefined();
    expect(component.state('user')).not.toHaveProperty('id');
    expect(component.state('poll')).toBeUndefined();
    expect(tutils.getSessionItem('users')).toBeNull();
    expect(component.state('showModal')).toBeDefined();
    expect(component.state('showModal')).toBeTruthy();
  });

  it('should not create state or session entries nor trigger modal on empty submit', async () => {
    const props = {
      state: {
        space: {
          id: 1,
          name: 'space1',
          status: true
        }
      }
    };

    const component = shallow(<Space location={props} />);
    const button = component.find('button');
    const spy = jest.spyOn(axios, 'get');

    await button.props().onClick();

    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenLastCalledWith(api.getPendingJoinPollInSpaceByName + '1/');
    expect(component.state('user')).toBeDefined();
    expect(component.state('user')).not.toHaveProperty('id');
    expect(component.state('poll')).toBeUndefined();
    expect(tutils.getSessionItem('users')).toBeNull();
    expect(component.state('showModal')).toBeDefined();
    expect(component.state('showModal')).toBeFalsy();
  });

  it('should call post api on modal confirm', () => {
    const props = {
      state: {
        space: {
          id: 1,
          name: 'space1',
          status: true
        }
      }
    };

    const component = shallow(<Space location={props} />);
    const confirmDialog = component.find('ConfirmDialog');
    const spy = jest.spyOn(axios, 'post');

    component.setState({
      user: {
        name: 'user1',
        space: 1
      }
    });
    confirmDialog.props().confirm();

    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenLastCalledWith(api.createNameRelatedPoll, {
      space: 1,
      name: 'user1'
    });
  });
});
