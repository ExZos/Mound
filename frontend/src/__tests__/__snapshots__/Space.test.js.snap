// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Space should render MessageSpace with props and session 1`] = `
<div
  id="space"
>
  <Sidebar />
  <WithStyles(ForwardRef(AppBar))
    position="sticky"
  >
    <WithStyles(ForwardRef(Toolbar))>
      <WithStyles(ForwardRef(IconButton))
        className="openMenu"
        onClick={[Function]}
        tabIndex="-1"
      >
        <MenuIcon />
      </WithStyles(ForwardRef(IconButton))>
      <WithStyles(ForwardRef(Menu))
        anchorOrigin={
          Object {
            "horizontal": "left",
            "vertical": "bottom",
          }
        }
        getContentAnchorEl={null}
        id="spaceNavMenu"
        onClose={[Function]}
        open={false}
        transformOrigin={
          Object {
            "horizontal": "left",
            "vertical": "top",
          }
        }
      >
        <WithStyles(ForwardRef(MenuItem))
          disabled={true}
        >
          user2
        </WithStyles(ForwardRef(MenuItem))>
        <WithStyles(ForwardRef(MenuItem))
          onClick={[Function]}
        >
          Messages
        </WithStyles(ForwardRef(MenuItem))>
        <WithStyles(ForwardRef(MenuItem))
          onClick={[Function]}
        >
          Polls
        </WithStyles(ForwardRef(MenuItem))>
        <WithStyles(ForwardRef(MenuItem))
          onClick={[Function]}
        >
          Change Name
        </WithStyles(ForwardRef(MenuItem))>
      </WithStyles(ForwardRef(Menu))>
      <WithStyles(ForwardRef(Typography))
        className="spaceName"
      >
        space2
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(IconButton))
        aria-label="close"
        className="closeSpace"
        onClick={[Function]}
        tabIndex="-1"
      >
        <CloseIcon />
      </WithStyles(ForwardRef(IconButton))>
    </WithStyles(ForwardRef(Toolbar))>
  </WithStyles(ForwardRef(AppBar))>
  <br />
  <MessageSpace
    updateState={[Function]}
    user={
      Object {
        "id": 2,
        "name": "user2",
        "poll": 2,
        "space": 2,
        "space_name": "space2",
        "space_status": true,
      }
    }
  />
</div>
`;

exports[`Space should render PendingSpace with props and session 1`] = `
<div
  id="space"
>
  <Sidebar />
  <WithStyles(ForwardRef(AppBar))
    position="sticky"
  >
    <WithStyles(ForwardRef(Toolbar))>
      <WithStyles(ForwardRef(IconButton))
        className="openMenu"
        onClick={[Function]}
        tabIndex="-1"
      >
        <MenuIcon />
      </WithStyles(ForwardRef(IconButton))>
      <WithStyles(ForwardRef(Menu))
        anchorOrigin={
          Object {
            "horizontal": "left",
            "vertical": "bottom",
          }
        }
        getContentAnchorEl={null}
        id="spaceNavMenu"
        onClose={[Function]}
        open={false}
        transformOrigin={
          Object {
            "horizontal": "left",
            "vertical": "top",
          }
        }
      >
        <WithStyles(ForwardRef(MenuItem))
          disabled={true}
        >
          userundefined
        </WithStyles(ForwardRef(MenuItem))>
        <WithStyles(ForwardRef(MenuItem))
          onClick={[Function]}
        >
          Messages
        </WithStyles(ForwardRef(MenuItem))>
        <WithStyles(ForwardRef(MenuItem))
          onClick={[Function]}
        >
          Polls
        </WithStyles(ForwardRef(MenuItem))>
        <WithStyles(ForwardRef(MenuItem))
          onClick={[Function]}
        >
          Change Name
        </WithStyles(ForwardRef(MenuItem))>
      </WithStyles(ForwardRef(Menu))>
      <WithStyles(ForwardRef(Typography))
        className="spaceName"
      >
        space3
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(IconButton))
        aria-label="close"
        className="closeSpace"
        onClick={[Function]}
        tabIndex="-1"
      >
        <CloseIcon />
      </WithStyles(ForwardRef(IconButton))>
    </WithStyles(ForwardRef(Toolbar))>
  </WithStyles(ForwardRef(AppBar))>
  <br />
  <PendingSpace
    updateState={[Function]}
    user={
      Object {
        "name": "userundefined",
        "poll": 3,
        "space": 3,
        "space_name": "space3",
        "space_status": false,
      }
    }
  />
</div>
`;

exports[`Space should render PendingUser with props and session  1`] = `
<div
  id="space"
>
  <Sidebar />
  <WithStyles(ForwardRef(AppBar))
    position="sticky"
  >
    <WithStyles(ForwardRef(Toolbar))>
      <WithStyles(ForwardRef(IconButton))
        className="openMenu"
        onClick={[Function]}
        tabIndex="-1"
      >
        <MenuIcon />
      </WithStyles(ForwardRef(IconButton))>
      <WithStyles(ForwardRef(Menu))
        anchorOrigin={
          Object {
            "horizontal": "left",
            "vertical": "bottom",
          }
        }
        getContentAnchorEl={null}
        id="spaceNavMenu"
        onClose={[Function]}
        open={false}
        transformOrigin={
          Object {
            "horizontal": "left",
            "vertical": "top",
          }
        }
      >
        <WithStyles(ForwardRef(MenuItem))
          disabled={true}
        >
          userundefined
        </WithStyles(ForwardRef(MenuItem))>
        <WithStyles(ForwardRef(MenuItem))
          onClick={[Function]}
        >
          Messages
        </WithStyles(ForwardRef(MenuItem))>
        <WithStyles(ForwardRef(MenuItem))
          onClick={[Function]}
        >
          Polls
        </WithStyles(ForwardRef(MenuItem))>
        <WithStyles(ForwardRef(MenuItem))
          onClick={[Function]}
        >
          Change Name
        </WithStyles(ForwardRef(MenuItem))>
      </WithStyles(ForwardRef(Menu))>
      <WithStyles(ForwardRef(Typography))
        className="spaceName"
      >
        space3
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(IconButton))
        aria-label="close"
        className="closeSpace"
        onClick={[Function]}
        tabIndex="-1"
      >
        <CloseIcon />
      </WithStyles(ForwardRef(IconButton))>
    </WithStyles(ForwardRef(Toolbar))>
  </WithStyles(ForwardRef(AppBar))>
  <br />
  <PendingUser
    updateState={[Function]}
    user={
      Object {
        "name": "userundefined",
        "poll": 3,
        "space": 3,
        "space_name": "space3",
        "space_status": true,
      }
    }
  />
</div>
`;

exports[`Space should render correctly with props but no session 1`] = `
<div
  id="space"
>
  <Sidebar />
  <WithStyles(ForwardRef(AppBar))
    position="sticky"
  >
    <WithStyles(ForwardRef(Toolbar))>
      <WithStyles(ForwardRef(Typography))
        className="spaceName"
      >
        space1
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(IconButton))
        aria-label="close"
        className="closeSpace"
        onClick={[Function]}
        tabIndex="-1"
      >
        <CloseIcon />
      </WithStyles(ForwardRef(IconButton))>
    </WithStyles(ForwardRef(Toolbar))>
  </WithStyles(ForwardRef(AppBar))>
  <br />
  <div>
    <form
      id="getUserInSpaceByName"
      onSubmit={[Function]}
    >
      <div
        className="textFieldWButton"
      >
        <WithStyles(ForwardRef(TextField))
          autoFocus={true}
          label="Type a user name..."
          name="name"
          onChange={[Function]}
          size="small"
          value=""
          variant="outlined"
        />
        <WithStyles(ForwardRef(Button))
          color="primary"
          onClick={[Function]}
          size="small"
          tabIndex="-1"
          type="submit"
          variant="outlined"
        >
          ENTER
        </WithStyles(ForwardRef(Button))>
      </div>
    </form>
    <ConfirmDialog
      confirm={[Function]}
      mBody="Do you want to request to join this space as ''"
      mHeader="User '' does not exist"
      showModal={false}
      toggleModal={[Function]}
    />
  </div>
</div>
`;
